Added Prettier as source code formatter

mqnguyen5 opened the pull request:
Fixes #24.

Current progress:

- [x] added prettier to dependencies
- [x] added prettier config file
- [x] added prettier ignore file to skip anything we don't want touched
- [x] changed npm scripts to include prettier
- [x] added git hook with Husky to run prettier on every commit
- [x] added a .vscode/ directory and include the prettier extension in the list of suggestions
- [ ] ran prettier over the entire code base so it gets formatted

mqnguyen5 commented:
This PR is ready for review, but I haven't finish all the tasks yet and need some feedbacks. I'm a bit unsure about the files/directory included inside `.prettierignore`. Can someone take a look and let me know what I'm missing?

Kevan-Y commented on .prettierignore:
```diff
@@ -0,0 +1,8 @@
+
+node_modules/
+static/
+
```
@mqnguyen5 I think you can also base on .gitignore. Like `/build` and `.docusaurus` you don't want those 2 folders to be reformated

mqnguyen5 replied:
Sounds good. I'll try to refer to the `.gitignore` to see what I can add. Thanks!

aserputov commented:
@mqnguyen5 LGTM!
[https://www.youtube.com/watch?v=rR4n-0KYeKQ](link)

mqnguyen5 commented:
I have included some changes, specifically:

- added a .vscode/ directory and include the prettier extension in the list of suggestions
- ran prettier over the entire code base so it gets formatted

Please have a look and let me know if I need to fix anything.

humphd commented on docs/A-Introduction/information.md:
```diff
@@ -8,115 +8,113 @@ sidebar_position: 2
-* Peta or P (=1024T): 1 Petabyte = 1024 * 1024 * 1024 * 1024 * 1024 bytes ~ 10<sup>15</sup> bytes
-* Exa or E (=1024P): 1 Exabyte = 1024 * 1024 * 1024 * 1024 * 1024 * 1024 bytes ~ 10<sup>18</sup> bytes
+- Kilo or k (=1024): 1 Kilobyte = 1024 bytes ~ 10<sup>3</sup> bytes
+- Mega or M (=1024k): 1 Megabyte = 1024 \* 1024 bytes ~ 10<sup>6</sup> bytes
```
I'm not sure what's happening here, this looks wrong

humphd replied:
Can you also update `CONTRIBUTING.md` to reflect that you need to have Prettier installed in VSCode, or run the prettier command to format?

mqnguyen5 replied:
From what I saw, apparently VSCode decided that using `-` is better than `*` for writing bulleted list.

As for updating `CONTRIBUTING.md`, I'll have a look and update it as soon as possible

humphd reviewed (commented):
Most of this looks good, but there are a few places that I don't understand what it's doing.

I also wonder if we should wait to do this until ***after*** all the other PRs land for 0.3, since this will be pretty disruptive to everyone's current work.

Kevan-Y commented on .vscode/extensions.json:
```diff
@@ -0,0 +1,8 @@
+{
+  "recommendations": [
+    "editorconfig.editorconfig",
+    "dbaeumer.vscode-eslint",
```
Do we need this extension?

humphd replied:
No, good catch.

mqnguyen5 replied:
For now I'm just following what we have on [Telescope](https://github.com/Seneca-CDOT/telescope/blob/master/.vscode/extensions.json). @humphd should I remove them?

humphd replied:
You can use Telescope as a guide, but this isn't a node/js project (e.g. we don't support the code, just the docs) so we don't need it here.

humphd commented on docs/A-Introduction/information.md:
```diff
@@ -8,115 +8,113 @@ sidebar_position: 2
-* Exa or E (=1024P): 1 Exabyte = 1024 * 1024 * 1024 * 1024 * 1024 * 1024 bytes ~ 10<sup>18</sup> bytes
+- Kilo or k (=1024): 1 Kilobyte = 1024 bytes ~ 10<sup>3</sup> bytes
+- Mega or M (=1024k): 1 Megabyte = 1024 \* 1024 bytes ~ 10<sup>6</sup> bytes
+- Giga or G (=1024M): 1 Gigabyte = 1024 _ 1024 _ 1024 bytes ~ 10<sup>9</sup> bytes
```
But look at this change, it's changed `*` to `_` which is semantically not the same thing.  I don't understand it.  I wonder if we need to embed this code in backticks to have it not get changed?

mqnguyen5 replied:
Somehow Prettier thought the `*` in `1024 * 1024 * 1024` as an indication for bulleted list or something and changed it into `_`. I can revert the formatting commit and embed this code in backticks to see if it helps.

mqnguyen5 replied:
I noticed that some lines used `\` to escape the `*` and some don't. I think that's why Prettier decided to change `1024 * 1024 * 1024` into `1024 _ 1024 _ 1024`.

mqnguyen5 replied:
I just reverted the formatting commit. I think we can either embed the line of code in backticks or escape it with `\`

mqnguyen5 commented:
For the Prettier formatting bug, I just reverted the formatting commit. I think we should double check the special characters used in the docs, specifically in `docs/A-Introduction/information.md`. We can either use `\` to escape the special characters, or wrap the lines of code in backticks.

For the extensions comment by @Kevan-Y and based on the suggestion of @humphd , I removed  `editorconfig.editorconfig` and `dbaeumer.vscode-eslint`.

Kevan-Y commented:
> For the Prettier formatting bug, I just reverted the formatting commit. I think we should double check the special characters used in the docs, specifically in `docs/A-Introduction/information.md`. We can either use `\` to escape the special characters, or wrap the lines of code in backticks.

In my opinion, I think the best option is to escape those characters. We will have to check each .md file and escape special characters.

> For the extensions comment by @Kevan-Y and based on the suggestion of @humphd , I removed editorconfig.editorconfig and dbaeumer.vscode-eslint.

Looks good. Thanks

mqnguyen5 commented:
> In my opinion, I think the best option is to escape those characters. We will have to check each .md file and escape special characters.

Good point. Since we have a lot of files that need formatting, I think we can file a separate issue for it, and merge what we have at the moment:

- added prettier to dependencies
- added prettier config file
- added prettier ignore file to skip anything we don't want touched
- changed npm scripts to include prettier
- added git hook with Husky to run prettier on every commit
- added a .vscode/ directory and include the prettier extension in the list of suggestions

@humphd what do you think about this?

humphd commented:
The minute we land the prettier stuff, we also need to format the whole tree at the same time.  We have to be ready to deal with the edge cases before this goes in.

I would file issues on the specific things that Prettier is breaking, get those fixed and merged, rebase this, and then run prettier over the whole tree and land.

mqnguyen5 commented:
Alright, I'll ran the docs against prettier again, and file issues explaining what to fix. Hopefully we can get it done by today

mqnguyen5 commented:
#56

mqnguyen5 commented:
@humphd I have filed an issue for the Prettier bug at #56. Hopefully we can get everyone to finish other PRs so we can merge this.

humphd commented:
A note that we'll circle back on this early next week once we've cleared the current PRs.

humphd commented:
@mqnguyen5 let's rebase this and re-run so we can review it again.

mqnguyen5 commented:
Will do.

humphd commented:
@mqnguyen5 we just merged the last of the PRs that were blocking this, do you want to rebase and try running it again to see how it is?  I think we can merge soon.

mqnguyen5 commented:
@humphd I just pull the latest changes and run them through Prettier.

humphd reviewed (approved):
This looks great, I have no concerns.  Well done.

Kevan-Y reviewed (approved):
LGTM

humphd commented:
@CameronGray1210 I'm going to merge this, and it will mean that you need to pull from the upstream repo and then re-run `yarn` to install the new dependencies (including prettier).  This change will automatically cause git to trigger the formatting when you commit anything new, so as long as people have run `yarn` before they `git commit`, it should remain consistent.

In a follow-up, I'll file an issue to add the `prettier-check` step to our CI for pull requests, so we can spot people forgetting to do this and get them to fix before we merge.

Also note: this change will recommend that people install the following VSCode extensions, which will help with the editing process, and run Prettier every time you save:

- https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode
- https://marketplace.visualstudio.com/items?itemName=streetsidesoftware.code-spell-checker

humphd commented:
https://github.com/Seneca-ICTOER/IPC144/issues/107#issuecomment-980040212
